File: solutions/python/best-time-to-buy-and-sell-stock.py
# id: 121
# title: Best Time to Buy and Sell Stock
# approach: Greedy (Track minimum price)
# time: O(n) | space: O(1)
# author: arnav-goel

import sys

def main():
    input = sys.stdin.read
    data = input().split()
    
    n = int(data[0])
    prices = list(map(int, data[1:n+1]))

    min_price = float('inf')
    max_profit = 0

    for price in prices:
        if price < min_price:
            min_price = price
        else:
            max_profit = max(max_profit, price - min_price)

    print(max_profit)

if __name__ == "__main__":
    main()

File: solutions/java/best-time-to-buy-and-sell-stock.java
// id: 121
// title: Best Time to Buy and Sell Stock
// approach: Greedy (Track minimum price)
// time: O(n) | space: O(1)
// author: arnav-goel

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int n = Integer.parseInt(br.readLine().trim());
        String[] input = br.readLine().trim().split(" ");
        
        int[] prices = new int[n];
        for (int i = 0; i < n; i++) {
            prices[i] = Integer.parseInt(input[i]);
        }

        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;

        for (int i = 0; i < n; i++) {
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            } else {
                maxProfit = Math.max(maxProfit, prices[i] - minPrice);
            }
        }

        System.out.println(maxProfit);
    }
}